remote-config/src/fetch/shared.rs

```diff
@@ -0,0 +1,266 @@
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
+use std::time::Duration;
+use tokio_util::sync::CancellationToken;
+use tracing::error;
+use crate::fetch::{ConfigFetcher, ConfigFetcherState, ConfigInvariants, FileStorage, OpaqueState};
+use crate::{RemoteConfigPath, Target};
+use tokio::time::sleep;
+use tokio::select;
+
+/// Fetcher which does a run-loop and carefully manages state around files, with the following
+/// guarantees:
+///  - A file at a given RemoteConfigPath will not be recreated as long as it exists
+///    I.e. it will always be drop()'ed before recreation.
+///  - It does not leak files which are no longer in use, i.e. it refcounts across all remote
+///    config clients sharing the same RefcountingStorage.
+///  - The state is always valid, i.e. there will be no intermittently expired files.
+///
+pub struct SharedFetcher {
+    /// (env, service, version) tuple representing the basic remote config target
+    pub target: Arc<Target>, // could be theoretically also Mutex<>ed if needed
+    /// A unique runtime id. It must not be used by any other remote config client at the same time.
+    /// Is allowed to be changed at any time.
+    pub runtime_id: Arc<Mutex<String>>,
+    /// Each fetcher must have an unique id. Defaults to a random UUID.
+    pub client_id: String,
+    cancellation: CancellationToken,
+    /// Interval used if the remote server does not specify a refetch interval
+    pub default_interval: AtomicU32,
+}
+
+pub struct FileRefcountData {
+    /// Primary refcounter:
+    ///  - When active (dropped_run_id == 0), the amount of runners holding it since the last
+    ///    remote config fetch.
+    ///  - When inactive (dropped_run_id > 0), the remaining amount of runners actively fetching
+    ///    remote config at the point in time dropped_run_id represents.
+    rc: AtomicU32,
+    /// 0, or point in time (see RunnersGeneration) where the file was moved to inactive.
+    dropped_run_id: AtomicU64,
+    pub path: RemoteConfigPath,
+    pub version: u64,
+}
+
+impl FileRefcountData {
+    pub fn new(version: u64, path: RemoteConfigPath) -> Self {
+        FileRefcountData {
+            rc: AtomicU32::new(0),
+            dropped_run_id: AtomicU64::new(0),
+            path,
+            version,
+        }
+    }
+}
+
+pub trait RefcountedFile {
```

Comment:
Can't this be achieved by using weak and strong references?

Reply:
No, using Weak references would imply that, once the fetchers no longer hold a reference to it, the underlying memory can no longer be accessed.  This may not be desired; refcounting allows careful control of only relinquishing it from fetchers, without affecting live references.

Reply:
I would give strong references to the shared clients and weak ones to everything else, so that it's self managed  and doesn't have `set` methods setting a counter